File: #0040.combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {

        vector<vector<int>> result;
        sort(candidates.begin(), candidates.end());
        vector<int> v;
        depthFirstSearch(candidates, target, 0, result, v);
        return result;
    }

private:
    void depthFirstSearch(const vector<int>& candidates,
                          int target, int start,
                          vector<vector<int>>& result,
                          vector<int>& v) {

        if (target == 0) {
            result.push_back(v);
            return;
        }

        int candidate;
        for (int i = start; i < candidates.size(); ++i) {
            candidate = candidates[i];
            if (candidate > target) {
                return;
            }

            if (i > start && candidates[i] == candidates[i - 1]) {
                continue;
            }
            v.push_back(candidate);
            depthFirstSearch(candidates, target - candidate, i + 1, result, v);
            v.pop_back();
        }
    }
};

File: #0040.combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    var buffer = [];
    var result = [];
    candidates.sort((a,b) => a - b);
    search(0, target);
    return result;

    function search(startIdx, target) {
        if (target < 0) {
            return;
        }

        if (target === 0) {
            return result.push(buffer.slice());
        }

        if (startIdx === candidates.length) {
            return;
        }

        for (let i = startIdx; i < candidates.length; i++) {
            if (startIdx < i
                && candidates[i] === candidates[i - 1]) {
                continue;
            }
            buffer.push(candidates[i]);
            search(i+1, target - candidates[i]);
            buffer.pop();

        }
    }
};
